resources/type-system/inference.md

```diff
@@ -1324,51 +1373,106 @@ determined as follows:
-
-  - Suspend the current invocation of the function body immediately enclosing
-    the await expression until after `o_2` completes.
-
-  - At some time after `o_2` is completed, control returns to the current
-    invocation.
-
-  - If `o_2` is completed with an error `x` and stack trace `t`, then `m`
-    completes with the exception `x` and stack trace `t`. _This trivially
-    satisfies expression soundess, since expression soundness is only concerned
-    with the value of the expression when it completes normally._
-
-  - If `o_2` is completed with an object `v`, then `m` completes with the value
-    `v`. _Since `o_2` is an instance of `Future<T>`, by future soundness, `v`
-    must be an instance of `T`. This satisfies expression soundness, since `v`
-    is an instance of `T` , and `T` is the static type._
+- If `T_1` is incompatible with await (as defined in the _extension types_
+  specification), then there is a compile-time error.
+
+- Let `T_2` be `flatten(T_1)`.
+
+- Let `m_2` be `@AWAIT_WITH_TYPE_CHECK<T_2>(m_1)`, with static type
+  `Future<T_2>`.
+
+  - _Note that in many circumstances, it will be trivial for the compiler to
+    establish that `m_1` always evaluates to an instance `v` that satisfies type
+    `Future<T_2>`, in which case `m_2` can be optimized to
+    `@PROMOTED_TYPE<Future<T_2>>(m_1)`._
+
+  - _For soundness, we must prove that whenever
+    `@AWAIT_WITH_TYPE_CHECK<T_2>(m_1)` evaluates `m_1` to a value `v` that is
+    __not__ an instance satisfying type `Future<T_2>`, that `v` __is__ an
+    instance satisfying type `T_2`. This is necessary to ensure that the future
+    created by `@AWAIT_WITH_TYPE_CHECK` will complete to a value satisfying its
+    type signature. Note that `v` is guaranteed to be an instance satisfying
+    type `T_1` (because `T_1` is the static type of `m_1`). So we can establish
+    soundness by assuming that `v` is an instance satisfying type `T_1` and not
+    an instance satisfying type `Future<T_2>`, and then considering two cases:_
+
+    - _If the runtime value of `v` is `null`, then by soundness, `T_1` must be
+      of the form `Null`, `dynamic`, `S*`, or `S?`. Considering each of these:_
+
+      - _If `T_1` is of the form `Null` or `dynamic`, then by the definition of
+        `flatten`, `T_2` must be the same as `T_1`. Therefore, `v` is an
+        instance satisfying type `T_2`, so soundness is satisfied._
+
+      - _If `T_1` is of the form `S*` or `S?`, then by the definition of
+        `flatten`, `T_2` must be of the form `flatten(S)*` or `flatten(S)?`,
+        respectively. `null` is an instance satisfying all types ending in `*`
+        and `?`, so soudness is satisfied._
+
+    - _Otherwise, we need to show that if `v` is a non-null instance satisfying
+      type `T_1`, but not an instance satisfying type `Future<T_2>`, then `v` is
+      an instance satisfying type `T_2`._
+
+    - _Substituting in the definition of `T_2`, we need to show that if `v` is a
```

Comment:
@fishythefish this sort of thing is the kind of coherence property that I can imagine pulling out as a lemma to be validated.  "Lemma: If `v` has static type `T` and runtime type `S` where `S` is not `Null` then either `S <: flatten(T)` or `S <: Future<flatten<T>>`.

Reply:
Is the "not `Null`" restriction important? If `S` is `Null` (i.e. `v` is `null`), then `T` must be nullable. Therefore, `flatten(T)` is nullable (this lemma can be proven separately if needed), so `S <: flatten(T)`.